Remove ImageService dependency and serve image blobs direct from BlobStore

anna-giang opened the pull request:
Previously, ImageService was used to generate a URL for the image in Blobstore, and this URL was stored in DataStore and also used to display the image. However, it is a known issue that ImageService does not work after the website has been deployed (see [[Step Internship] Known Issues and FAQs](https://docs.google.com/document/d/1_Hyte0_3IJ9nPglQ4G116BXjuIK3dXxI5tQXy9rPpp0/)). 
Hence, the ImageService dependency is removed here, and instead the image blobs are served to the client code directly from Blobstore.

anna-giang committed:
New ImageBlobsServlet serving image blobs directly from BlobStore

Images will be retrieved using a GET request with a blobkey via this servlet rather than using ImageService.

anna-giang committed:
Add image blobkey to response from DataServlet and script.js to fetch images using blobkey

DataServlet no longer uses ImageService to generate a URL for the image, and instead returns the BlobKey string of the image in its response.
script.js will pass this blobkey to fetchCommentImage to fetch the image, and generate a URL for it. This URL will then be used to display the image with the associated comment.
In this way, the ImageService dependency that does not work in production is removed.

anna-giang committed:
Change comment image width in display

austinct commented on portfolio/src/main/webapp/script.js:
```diff
@@ -91,20 +93,24 @@ function fetchComments(quantity=5) {
-              + imageUrl + '></a></div></div>'; // add image
-        }
+        let blobKey = commentData[i].blobKey;
+        fetchCommentImage(blobKey).then((url) => {
```
An idea to make this nicer - you could turn fetchComments into an `async` function. Then you don't need this double nested `then()`. You don't have to do this in this pull request though.

anna-giang replied:
Yep, I'll keep this in mind and come back to it if I have time after week 5's content! 😊

austinct commented on portfolio/src/main/java/com/google/sps/servlets/ImageBlobsServlet.java:
```diff
@@ -0,0 +1,39 @@
+    if (queryString == null) {
+      return;
+    }
+    BlobKey blobKey = new BlobKey(request.getParameter("blob-key"));
```
Have you thought about what happens if the user gives an invalid blob-key, or doesn't provide one?

anna-giang replied:
I did think about these cases before, but I knew that my current front-end code checks for whether there is a BlobKey before passing it to the server (hence there will be no null BlobKeys being passed to server). And I knew that invalid BlobKeys will be a 404 error, and to the user it will just look like there wasn't an image submitted in the first place (of course, with errors in the console).

But I guess it's best practice to handle edge cases because I don't want unexpected behaviour later on if more code was to use this servlet 😅 so I've handled those cases now

austinct replied:
Thanks for handling!

anna-giang committed:
Rename url to imageUrl

url was assigned to the variable imageUrl before it was used anyway, so no need for the two names.

anna-giang committed:
Handle blob-key exceptional cases

When invalid blobkey is submitted to the /serve-images server, the server will respond with a 404 (image not found), which is handled in the front end by displaying "image not found" to the user, and console logging the error.
When no blobkey is provided, the server returns will return with null.

austinct commented on portfolio/src/main/webapp/script.js:
```diff
@@ -183,9 +185,16 @@ async function fetchCommentImage(blobKey) {
+    const imageURL = window.URL.createObjectURL(blob);
+    return imageURL; 
+  }
+  else {
```
`} else {`

austinct commented on portfolio/src/main/webapp/script.js:
```diff
@@ -99,8 +99,10 @@ function fetchComments(quantity=5) {
+          if (imageUrl === null) { // No image
             commentContent += '</div>'; // close outer div
+          } else if (imageUrl === undefined) { // Error occured
+            commentContent += '<div class="body-text">Error fetching image</div></div>';
```
General comment: rather than handcrafting html with string manipulations, consider using `document.createElement()`, `element.appendChild()`, etc. to add elements to the page. No need to fix it here, but something to consider for the future.

anna-giang replied:
Yep, Eizan has also mentioned this to me previously, and I have tried it out for a previous version of the comments output, but it has so many more lines of code, so I thought it was overkill:

```
        var commentDiv = document.createElement('div'); // div to store one comment
        commentDiv.className = "comment";

        let commentText = commentData[i].commentText; 
        // Comment author name: "'Anonymous' if there was not a name submitted."
        let commentAuthor = commentData[i].commentAuthor === "" ? "Anonymous" : commentData[i].commentAuthor; 
        let imageUrl = commentData[i].attachedImage;

        let textContainer = document.createElement('div'); // Container for comment text & author name
        textContainer.className = "flex-item";
        
        // Author name <p> element
        let authorP = document.createElement('p');
        authorP.className = "body-text";
        let bolding = document.createElement('b');
        bolding.innerText = commentAuthor;
        authorP.appendChild(bolding);

        // Comment text <p> element
        let commentP = document.createElement('p');
        commentP.className = "body-text";
        commentP.innerText = commentText;

        textContainer.appendChild(authorP);
        textContainer.appendChild(commentP);
        commentDiv.appendChild(textContainer);

        // Display image as well, if there is an image
        if (imageUrl) {
          let imageContainer = document.createElement('div');
          imageContainer.className = "flex-item";
          
          let imageLink = document.createElement('a');
          imageLink.href = imageUrl;
          imageLink.target = "_blank"; // link will open in new tab
          
          let image = document.createElement('img');
          image.className = "comment-image";
          image.src = imageUrl;

          imageLink.appendChild(image);
          imageContainer.appendChild(imageLink);
          commentDiv.appendChild(imageContainer);
        }
        document.getElementById('comment-list').appendChild(commentDiv);
      }
```

I don't know if there is a way to shorten this? 😅

austinct replied:
Yeah ok, this is still hard to read haha.

The main things in the original code we want to do is limit string manipulations (e.g. try avoid all the `+=`s to append to the html string) and have all the HTML formatted and in as few assignments as possible so it's easy to see the structure. So, try do as much as possible in each innerHTML string assignment, and then edit it using element manipulation functions.

E.g below is a middle ground:

```js
let imageContent = "";
if (imageUrl === undefined) {
  imageContent = '<div class="body-text">Error fetching image</div>';
} else if (imageUrl !== null) {
  imageContent = `
    <div class="flex-item">
      <a href="${imageUrl}" target="_blank">
        <img class="comment-image" src="${imageUrl}">
      </a>
    </div>`;
}

var commentDiv = document.createElement('div'); // div to store one comment
commentDiv.className = "comment";
commentDiv.innerHTML = `
  <div class="flex-item">
    <p class="body-text"><b>${commentAuthor}</b></p>
    ${authorEmailContent}
    <p class="body-text">${commentText}</p>
  </div>
  ${imageContent}`
document.getElementById('comment-list').appendChild(commentDiv);
```

Although it still has some issues:
- User comments are inserted directly as html - generally you want to use `element.textContent = comment` to insert user data as they could contain html tags.
- I don't like how imageContent is done - I'd rather create it as an element and then `commentDiv.appendChild(imageContent)`

austinct replied:
Also long isn't necessarily bad, if making it longer makes it more readable, that's acceptable.
